src/common/snippets/src/lowered/pass/define_buffer_clusters.cpp

```diff
@@ -42,141 +69,126 @@ size_t DefineBufferClusters::get_cluster_buffer_id(const BufferCluster& cluster)
-                const auto output_params =
-                    ShiftPtrParams(data_sizes[output_buffer_port_idx], ptr_increments[output_buffer_port_idx], final_offsets[output_buffer_port_idx]);
-
-                // If data pointer shift parameters are unknown on model compilation stage (dynamic),
-                // we cannot be sure that these data pointers will be proportionally shifted in runtime.
-                if (input_params.is_static() && output_params.is_static() && input_params == output_params) {
-                    const auto cluster_it = find_cluster_by_expr(input_buffer_expr);
-                    OPENVINO_ASSERT(cluster_it != m_clusters.end(), "Buffer on inputs of Loop must be already saved in clusters");
-                    // Add to the existing cluster
-                    has_been_added = cluster_it->insert(output_buffer_expr).second;
-                    OPENVINO_ASSERT(has_been_added, "Buffer has not been saved in cluster");
-                    // Remove input buffer because we have already use its memory
-                    visited_buffers.insert(input_buffer_expr);
-                    break;
-                }
+            const auto in_path = MarkInvariantShapePath::getInvariantPortShapePath(*input_buffer_port_info.port.expr_port);
+            const auto out_path = MarkInvariantShapePath::getInvariantPortShapePath(*output_buffer_port_info.port.expr_port);
+
+            //  - Memory can be shared if Buffer has the same allocation size.
+            if (input_buffer_expr->is_defined() && output_buffer_expr->is_defined()) {
+                if (input_buffer_expr->get_allocation_size() != output_buffer_expr->get_allocation_size())
+                    continue;
+            } else {
+                // If allocation sizes are undefined, we can check if they have the same allocation sizes in runtime:
+                //  - they should calculate allocation size using the common algorithm from `BufferExpression::init_allocation_size`.
+                //  - they should be on the same `shape path` - this guarantees the same shapes in runtime.
+                const auto is_the_same_init_algo =
+                    utils::everyone_is(BufferExpression::get_type_info_static(), input_buffer_expr->get_type_info(), output_buffer_expr->get_type_info());
+                const auto is_the_same_shape_paths = in_path == out_path;
+                if (!is_the_same_init_algo || !is_the_same_shape_paths)
+                    continue;
             }
-            if (has_been_added) break;
+
+            //  - Memory can be reused if there are the same LoopPortDesc (data size, final offsets, ptr increments).
+            //    LoopPorts should be on the same shape-path and have the same value of `is_incremented`.
+            if (in_path != out_path || input_buffer_port_info.port.is_incremented != output_buffer_port_info.port.is_incremented)
+                continue;
```

Comment:
Minor: can we move this check upper, right after `in_path`/`out_path` initialization? I suppose this easy and obvious check can be done before all others (which depend on buffer state - defined or not)

Also, we can remove `is_the_same_shape_paths` check on L147, since it partially duplicates this one